src/interpreter_funcs/voxel_boolean_difference.rs

```diff
@@ -0,0 +1,262 @@
+use std::error;
+use std::f32;
+use std::fmt;
+use std::ops::Bound;
+use std::sync::Arc;
+
+use nalgebra::Vector3;
+
+use crate::analytics;
+use crate::interpreter::{
+    BooleanParamRefinement, Float3ParamRefinement, Func, FuncError, FuncFlags, FuncInfo,
+    LogMessage, ParamInfo, ParamRefinement, Ty, UintParamRefinement, Value,
+};
+use crate::mesh::voxel_cloud::{self, ScalarField};
+
+const VOXEL_COUNT_THRESHOLD: u32 = 50000;
+
+#[derive(Debug, PartialEq)]
+pub enum FuncBooleanDifferenceError {
+    WeldFailed,
+    EmptyScalarField,
+    VoxelDimensionsZero,
+    TooManyVoxels(u32, f32, f32, f32),
+}
+
+impl fmt::Display for FuncBooleanDifferenceError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            FuncBooleanDifferenceError::WeldFailed => write!(
+                f,
+                "Welding of separate voxels failed due to high welding proximity tolerance"
+            ),
+            FuncBooleanDifferenceError::EmptyScalarField => write!(f, "The resulting scalar field is empty"),
+            FuncBooleanDifferenceError::VoxelDimensionsZero => write!(f, "One or more voxel dimensions are zero"),
+            FuncBooleanDifferenceError::TooManyVoxels(max_count, x, y, z) => write!(
+                f,
+                "Too many voxels. Limit set to {}. Try setting voxel size to [{:.3}, {:.3}, {:.3}] or more.",
+                max_count, x, y, z
+            ),
+        }
+    }
+}
+
+impl error::Error for FuncBooleanDifferenceError {}
+
+pub struct FuncBooleanDifference;
+
+impl Func for FuncBooleanDifference {
+    fn info(&self) -> &FuncInfo {
+        &FuncInfo {
+            name: "Difference",
+            description: "BOOLEAN DIFFERENCE OF VOXEL CLOUDS FROM TWO MESH GEOMETRIES\n\
+            \n\
+            Converts the input mesh geometries into voxel clouds, then performs \
+            boolean difference of the first to second voxel clouds and eventually \
+            materializes the resulting voxel cloud into a welded mesh. \
+            Boolean difference removes volume of the first geometry where there is \
+            a volume in the second geometry (removes the second volume from the first one). \
+            It is equivalent to arithmetic subtraction.\n\
```

Comment:
Not equivalent, but analogous.

Reply:
are you sure?

Reply:
No, I am not. But it feels more correct. Equivalence is quite a strong term.